### PR TITLE
inherited microscope in configuration

### PR DESCRIPTION
**Pull request by annie-xd-wang**
In the configuration.yaml

microscopes:
    Microscope-1:
        camera:
            ....
        daq:
            ....
    Microscope-2 (Microscope-1):
       ....

### CODE REVIEW

File: src/navigate/config/config.py [hunk 1]
```diff
@@ -913,19 +913,64 @@ def verify_configuration(manager, configuration):
     
     Supports old version of configurations.
     """
+    device_config = configuration["configuration"]["microscopes"]
+
+    # get microscope inheritance sequence
+    microscope_name_seq = []
+    inherited_microscope_dict = {}
+    for microscope_name in device_config.keys():
```
Comment (conorhughmcfadden) [inline src/navigate/config/config.py]: If I try a toy example in a jupyter notebook of lines 916-953, defining `device_config` as,

```python
device_config = {
    "ScopeA": {
        "Dev_A1": {}
    },
    "ScopeB(ScopeA)": {
    }
}
```

I get a RuntimeError:

image

I can solve it by changing line 921 to,

```python
microscope_names_list = list(device_config.keys())
for microscope_name in microscope_names_list:
```

Interesting that this is not an issue when running Navigate. Maybe the error is somehow ignored?

**Review by conorhughmcfadden — COMMENTED** [review]
Looks and works great! Even works in a toy example when you define multiple base microscopes:

```python
device_config = {
    "ScopeA": {
        "Dev_A1": {}
    },
    "ScopeB(ScopeA)": {
    },
    "ScopeC": {
        "Dev_C1": {}
    }
}

microscope_name_seq = []
inherited_microscope_dict = {}
microscope_names_list = list(device_config.keys())
for microscope_name in microscope_names_list:
    try:
        parenthesis_l = microscope_name.index("(")
    except ValueError:
        if microscope_name.strip() not in microscope_name_seq:
            microscope_name_seq.append(microscope_name.strip())
        continue

    if ")" not in microscope_name[parenthesis_l+1:] :
        microscope_name_seq.append(microscope_name.strip())
        continue

    parenthesis_r = microscope_name[parenthesis_l+1:].index(")")
    parent_microscope_name = microscope_name[parenthesis_l+1: parenthesis_l+parenthesis_r+1].strip()

    if parent_microscope_name not in microscope_name_seq:
        microscope_name_seq.append(parent_microscope_name)

    idx = microscope_name_seq.index(parent_microscope_name)
    child_microscope_name = microscope_name[:parenthesis_l].strip()
    microscope_name_seq.insert(idx+1, child_microscope_name)
    inherited_microscope_dict[child_microscope_name] = parent_microscope_name
    device_config[child_microscope_name] = device_config.pop(microscope_name)

# update microscope devices from parent microscope
for microscope_name in microscope_name_seq:
    if microscope_name not in inherited_microscope_dict:
        continue
    parent_microscope_name = inherited_microscope_dict[microscope_name]
    for device_name in device_config[parent_microscope_name].keys():
        if device_name not in device_config[microscope_name].keys():
            device_config[microscope_name][device_name] = device_config[parent_microscope_name][device_name]

print(device_config)

```
Output:
{'ScopeA': {'Dev_A1': {}}, 'ScopeC': {'Dev_C1': {}}, 'ScopeB': {'Dev_A1': {}}}
```

Just one small comment included involving overwriting the device_dict keys during the for loop... Again, navigate ran fine so maybe it's OK. But maybe dumping the microscope names in a list and then running the for loop off of that is safer?

**Comment by AdvancedImagingUTSW** [comment]
Thanks for doing such a careful evaluation, @conorhughmcfadden.  We can see what @annie-xd-wang has to say about the things you brought up.

File: src/navigate/config/config.py (hunk 1 above)
Comment (annie-xd-wang) [reply src/navigate/config/config.py]: Nice finding. We should take your solution.

**Review by conorhughmcfadden — COMMENTED** [review]
OK, looks good to me! - Conor

**Comment by AdvancedImagingUTSW** [comment]
@conorhughmcfadden - Go ahead and submit PR approval and merge it.

**Review by conorhughmcfadden — APPROVED** [review]
Approved!

**Pull request merged**